base/uk.ac.stfc.isis.ibex.opis/check_opi_format.py

```diff
@@ -0,0 +1,270 @@
+import os
+import sys
+from lxml import etree
+from time import gmtime, strftime
+
+
+class CheckOpiFormat:
+
+    # Directory to iterate through
+    root_directory = r"C:\Instrument\Dev\ibex_gui\base\uk.ac.stfc.isis.ibex.opis\resources"
+
+    # Files ending with .opi are parsed:
+    file_extension = r".opi"
+
+    # If a word contains any of the following, the whole word will be ignored
+    ignore = \
+        ["$", "&", "#", '"', "'", "(", ")", "OPI", "PSU", "HW", "Hz", "LED", "A:", "B:", "C:", "D:"]
+
+    # When checking capitalisation will ignore words <= this length
+    # This is to avoid
+    ignore_short_words_limit = 3
+
+    # Start of the XPATH for a widget
+    widget_xpath = "widget[@typeId='org.csstudio.opibuilder.widgets."
+
+    def __init__(self):
+        self.errors = []
+
+    @staticmethod
+    def get_tree(filepath):
+        with open(filepath) as opi_file:
+            parser = etree.XMLParser(remove_blank_text=True)
+            return etree.parse(opi_file, parser)
+
+    def file_iterator(self):
+        for filename in os.listdir(self.root_directory):
+            if filename.endswith(self.file_extension):
+                filepath = os.path.join(self.root_directory, filename)
+                yield filepath
+
+    def check_condition(self, root, xpath, error_message):
+        for error in root.xpath(xpath):
+            err = "Error on line " + str(error.sourceline) + ": " + etree.tostring(error) \
+                + "\n... " + error_message
+            self.errors.append(err)
+
+    def check_title_case(self, root, xpath, error_message):
+        for name in root.xpath(xpath):
+            capitalisation_error = False
+            words = name.text.split()
+
+            # Ignore words less than 4 characters as these are probably prepositions
+            for word in words:
+                if len(word) > self.ignore_short_words_limit \
+                        and word.title() != word \
+                        and not any(s in word for s in self.ignore):
+                    capitalisation_error = True
+
+            if capitalisation_error:
+                err = "Error on line " + str(name.sourceline) + ": " + etree.tostring(name) \
+                      + "\n... " + error_message
+                self.errors.append(err)
+
+    def check_sentence_case(self, root, xpath, error_message):
+
+        for name in root.xpath(xpath):
+            capitalisation_error = False
+            words = name.text.split()
+
+            # Ignore words less than 4 characters as these are probably prepositions
+            for word in words:
+                if len(word) > self.ignore_short_words_limit \
+                        and not words[0] == words[0].title() \
+                        and not word.lower() == word \
+                        and not any(s in word for s in self.ignore):
```

Comment:
The functionality is currently broken, as this essentially only checks the first word. As long as that's in title case(`and not words[0] == words[0].title()`), the if-check will always fail (thus, no error)